File: Batch7_UdayKiranKomara_EmployeeManagementSystem_/EMS.API/Services/EmployeeService.cs
using EMS.API.DTOs;
using EMS.API.Models;
using EMS.API.Repositories;
using Microsoft.EntityFrameworkCore;

namespace EMS.API.Services;

public class EmployeeService
{
    private readonly IEmployeeRepository _repo;

    public EmployeeService(IEmployeeRepository repo)
    {
        _repo = repo;
    }

    public async Task<PagedResult<EmployeeResponseDto>> GetAllAsync(EmployeeQueryParams query)
    {
        var q = _repo.GetQueryable();

        // SEARCH
        if (!string.IsNullOrEmpty(query.Search))
        {
            var term = query.Search.ToLower();
            q = q.Where(e =>
                (e.FirstName + " " + e.LastName).ToLower().Contains(term) ||
                e.Email.ToLower().Contains(term));
        }

        // FILTER
        if (!string.IsNullOrEmpty(query.Department))
            q = q.Where(e => e.Department == query.Department);

        if (!string.IsNullOrEmpty(query.Status))
            q = q.Where(e => e.Status == query.Status);

        // SORT
        q = query.SortBy switch
        {
            "salary" => query.SortDir == "asc"
                ? q.OrderBy(e => e.Salary)
                : q.OrderByDescending(e => e.Salary),

            "joinDate" => query.SortDir == "asc"
                ? q.OrderBy(e => e.JoinDate)
                : q.OrderByDescending(e => e.JoinDate),

            _ => query.SortDir == "asc"
                ? q.OrderBy(e => e.LastName).ThenBy(e => e.FirstName)
                : q.OrderByDescending(e => e.LastName).ThenByDescending(e => e.FirstName)
        };

        var total = await q.CountAsync();

        var data = await q
            .Skip((query.Page - 1) * query.PageSize)
            .Take(query.PageSize)
            .Select(e => new EmployeeResponseDto
            {
                Id = e.Id,
                FullName = e.FirstName + " " + e.LastName,
                Email = e.Email,
                Department = e.Department,
                Designation = e.Designation,
                Salary = e.Salary,
                JoinDate = e.JoinDate,
                Status = e.Status
            })
            .ToListAsync();

        return new PagedResult<EmployeeResponseDto>
        {
            Data = data,
            TotalCount = total,
            Page = query.Page,
            PageSize = query.PageSize,
            TotalPages = (int)Math.Ceiling(total / (double)query.PageSize),
            HasNextPage = query.Page * query.PageSize < total,
            HasPrevPage = query.Page > 1
        };
    }

    public async Task<EmployeeResponseDto?> GetByIdAsync(int id)
    {
        var e = await _repo.GetByIdAsync(id);
        if (e == null) return null;

        return new EmployeeResponseDto
        {
            Id = e.Id,
            FullName = e.FirstName + " " + e.LastName,
            Email = e.Email,
            Department = e.Department,
            Designation = e.Designation,
            Salary = e.Salary,
            JoinDate = e.JoinDate,
            Status = e.Status
        };
    }

    public async Task<bool> CreateAsync(EmployeeRequestDto dto)
    {
        if (await _repo.EmailExistsAsync(dto.Email))
            return false;

        var emp = new Employee
        {
            FirstName = dto.FirstName,
            LastName = dto.LastName,
            Email = dto.Email,
            Phone = dto.Phone,
            Department = dto.Department,
            Designation = dto.Designation,
            Salary = dto.Salary,
            JoinDate = dto.JoinDate,
            Status = dto.Status
        };

        await _repo.AddAsync(emp);
        return true;
    }

    public async Task<bool> UpdateAsync(int id, EmployeeRequestDto dto)
    {
        var emp = await _repo.GetByIdAsync(id);
        if (emp == null) return false;

        if (await _repo.EmailExistsAsync(dto.Email, id))
            return false;

        emp.FirstName = dto.FirstName;
        emp.LastName = dto.LastName;
        emp.Email = dto.Email;
        emp.Phone = dto.Phone;
        emp.Department = dto.Department;
        emp.Designation = dto.Designation;
        emp.Salary = dto.Salary;
        emp.JoinDate = dto.JoinDate;
        emp.Status = dto.Status;
        emp.UpdatedAt = DateTime.UtcNow;

        await _repo.UpdateAsync(emp);
        return true;
    }

    public async Task<bool> DeleteAsync(int id)
    {
        var emp = await _repo.GetByIdAsync(id);
        if (emp == null) return false;

        await _repo.DeleteAsync(emp);
        return true;
    }
}

File: Batch7_UdayKiranKomara_EmployeeManagementSystem/js/employeeService.js
const employeeService = (() => {

    function getAll() {
        return storageService.getAll();
    }

    function getById(id) {
        return storageService.getById(id);
    }

    function add(data) {
        const newEmployee = {
            id: storageService.nextId(),
            firstName: data.firstName,
            lastName: data.lastName,
            email: data.email,
            phone: data.phone,
            department: data.department,
            designation: data.designation,
            salary: Number(data.salary),
            joinDate: data.joinDate,
            status: data.status
        };

        storageService.add(newEmployee);
        return newEmployee;
    }

    function update(id, data) {
        storageService.update(id, data);
    }

    function remove(id) {
        storageService.remove(id);
    }

    function search(query) {
        const employees = storageService.getAll();

        if (!query) return employees;

        const lowerQuery = query.toLowerCase();

        return employees.filter(emp => {
            const fullName = `${emp.firstName} ${emp.lastName}`.toLowerCase();
            const email = emp.email.toLowerCase();

            return fullName.includes(lowerQuery) || email.includes(lowerQuery);
        });
    }

    function filterByDepartment(dept) {
        const employees = storageService.getAll();

        if (!dept) return employees;

        return employees.filter(emp => emp.department === dept);
    }

    function filterByStatus(status) {
        const employees = storageService.getAll();

        if (!status) return employees;

        return employees.filter(emp => emp.status === status);
    }

    function applyFilters(searchQuery, dept, status) {
        let employees = storageService.getAll();

        if (searchQuery) {
            const lower = searchQuery.toLowerCase();

            employees = employees.filter(emp => {
                const name = `${emp.firstName} ${emp.lastName}`.toLowerCase();
                return name.includes(lower) || emp.email.toLowerCase().includes(lower);
            });
        }

        if (dept) {
            employees = employees.filter(emp => emp.department === dept);
        }

        if (status) {
            employees = employees.filter(emp => emp.status === status);
        }

        return employees;
    }

    function sortBy(field, direction = "asc") {
        const employees = [...storageService.getAll()];

        employees.sort((a, b) => {

            let valA = a[field];
            let valB = b[field];

            if (field === "name") {
                valA = a.lastName.toLowerCase();
                valB = b.lastName.toLowerCase();
            }

            if (field === "joinDate") {
                valA = new Date(a.joinDate);
                valB = new Date(b.joinDate);
            }

            if (valA < valB) return direction === "asc" ? -1 : 1;
            if (valA > valB) return direction === "asc" ? 1 : -1;

            return 0;
        });

        return employees;
    }

    return {
        getAll,
        getById,
        add,
        update,
        remove,
        search,
        filterByDepartment,
        filterByStatus,
        applyFilters,
        sortBy
    };

})();